### PR TITLE
Use x-macros for action args too

### LINKED ISSUE #3475: Automatically generate `ShortcutAction` code from ~JSON Schema~ X-Macros
While working on #3391, I felt like there was a ton of boilerplate that was needed for each and every keybindings. The thought crossed my mind that we could do better, but I wasn't sure how. In my mind, I imagined a single file that described each of the `ShortcutAction`s and their possible args, and a small script that just automatically generated all the necessary code to parse them, their args, and hook them up to the `TerminalApp`. 

Now that I've thought on it more, here are some of the steps that'll be needed:

* [ ] Take code from my command palette branch for `ShortcutActionDispatch`. This is a class that just handles the dispatching of actions. In that branch, I needed it because both keybindings and commands needed to be able to dispatch `ShortcutAction`s, but only `AppKeyBindings` needed to do a `KeyChord` lookup. This code being encapsulated will be necessary for a future step.
* [ ] #2550 Add more templated json utils. We'll need this to be able to automagically generate the code to parse a given type from a `Json::Value`
* [ ] Move the code around so that anything having to do with the entirety of the list of `ShortcutAction`s is in single files by themselves:
  - [ ] AppKeybindingSerialization should only deal with parsing keychords. Lets put the `ShortcutAction`s and their keys in another file
  - [ ] Add a `IShortcutActionHandler` interface, with `_Handle{{SomeShortcutAction}}(auto&& sender, ActionArgs args)` methods for each `ShortcutAction`. It should declare all the event handlers, but not the definitions.
  - [ ] Make `TerminalPage` extend `IShortcutActionHandler`. This will be necessary to make sure TerminalPage implements each of them.
  - [ ] That one method that hooks up the `TerminalPage` to the `ShortcutActionDispatch`. Pull that into it's own file as well.
* [ ] Write a script that parses `profiles.schema.json` for each of the "*ShortcutAction" definitions, and turns them into code in all the necessary files
  - [ ] `ShortcutActionKeys.h`
  - [ ] `ActionAndArgs.idl`/`.h`/`.cpp`
  - [ ] `ShortcutActionDispatch.idl`/`.h`/`.cpp`
  - [ ] `IShortcutActionHandler.h`
  - [ ] For actions that are listed in `ShortcutAction`s but don't have a proper `Args` type, then make sure that they're still dispatched correctly
  - [ ] When the number of args _changes_, write the above files, and display a message that the developer needs to update `AppActionHandlers.cpp` to add implementations for the new types.
  - [ ] Validate that all "*ShortcutAction"s listed in the schema actually show up in the `onOf` schema for `Keybinding`

Each top-level box could be it's own task here.

In the end, the developer should be able to modify the schema for add their new argument or `ShortcutAction`, and run a script, and the script should automatically write all the modifications necessary.

I'll probably need to do this in powershell. `cmd` would be impossible, and as much as I love `python`, I don't think that's really a part of our toolchain at all currently.

### PR DESCRIPTION
**Pull request by zadjii-msft**
This adds x-macros for each of the actions, greatly reducing the amount of boilerplate needed for each action args. 

Originally, I wanted to do more with this, but I think the x-macros we've discovered properly treads the line of ease-of-use and native c++ support, with how much it'll do for us

* [x] Closes #3475
* [x] Sure enough, the tests still pass.
* [ ] mmmmmmm  image

### CODE REVIEW

File: src/cascadia/TerminalSettingsModel/ActionArgsMagic.h [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+// expanded. Pretty critical for tracking down extraneous commas, etc.
+
+// Property definitions, and JSON keys
+#define DECLARE_ARGS(type, name, jsonKey, required, ...)    \
+    static constexpr std::string_view name##Key{ jsonKey }; \
+    ACTION_ARG(type, name, ##__VA_ARGS__);
+
+// Parameters to the non-default ctor
+#define CTOR_PARAMS(type, name, jsonKey, required, ...) \
+    const type &name##Param,
```
Comment (lhecker) [inline src/cascadia/TerminalSettingsModel/ActionArgsMagic.h:61]: Why don't you just remove the trailing comma here? Then you won't need `InitListPlaceholder` either.

File: src/cascadia/TerminalSettingsModel/ActionArgsMagic.h [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+
+#define ACTION_ARG_BODY(className, argsMacro)                                   \
+    className() = default;                                                      \
+    className(                                                                  \
+        argsMacro(CTOR_PARAMS) InitListPlaceholder = {}) :                      \
+        argsMacro(CTOR_INIT) _placeholder{} {};                                 \
+    argsMacro(DECLARE_ARGS);                                                    \
+                                                                                \
+private:                                                                        \
+    InitListPlaceholder _placeholder;                                           \
```
Comment (lhecker) [inline src/cascadia/TerminalSettingsModel/ActionArgsMagic.h:120]: You don't need to store the placeholder.

File: src/cascadia/TerminalSettingsModel/ActionArgsMagic.h [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+// Use ACTION_ARG_BODY when you've got some other methods to add to the args class.
+// case in point:
+//   * NewTerminalArgs has a ToCommandline method it needs to additionally declare.
+//   * GlobalSummonArgs has the QuakeModeFromJson helper
+
+#define ACTION_ARG_BODY(className, argsMacro)                                   \
+    className() = default;                                                      \
+    className(                                                                  \
+        argsMacro(CTOR_PARAMS) InitListPlaceholder = {}) :                      \
+        argsMacro(CTOR_INIT) _placeholder{} {};                                 \
```
Comment (lhecker) [inline src/cascadia/TerminalSettingsModel/ActionArgsMagic.h:116]: This `;` isn't needed.

**Review by lhecker — APPROVED** [review]

File: src/cascadia/TerminalSettingsModel/ActionArgsMagic.h (hunk 2 above)
Comment (lhecker) [reply src/cascadia/TerminalSettingsModel/ActionArgsMagic.h:120]: `[[no_unique_address]]` will be helpful here in a few months with C++20.

Comment (miniksa) [reply src/cascadia/TerminalSettingsModel/ActionArgsMagic.h:120]: > few months

Wow you're optimistic, @lhecker.

**Review by carlos-zamora — APPROVED** [review]

File: src/cascadia/TerminalSettingsModel/ActionArgsMagic.h [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+struct InitListPlaceholder
+{
+};
+
+//
+// The complete ActionAndArgs definition. Each macro above ACTION_ARGS_STRUCT is
+// some element of the class definition that will use the x-macro.
+//
+// You'll author a new arg by:
+//   1: define a new x-macro above with all it's properties
```
Comment (miniksa) [inline src/cascadia/TerminalSettingsModel/ActionArgsMagic.h:42]: "Above".... probably should be more specific as to where.

**Review by miniksa — APPROVED** [review]
Yeah that's some magic alright.

**Pull request merged**